### ISSUE TITLE
ManyToOne lazy association is not loaded with Mutiny.fetch method 

### ISSUE THREAD

**Issue by Davide-Di-Do** [hibernate/hibernate-reactive#975]
Hi,
as mentioned in the title of the issue, when you try to load with Mutiny.fetch method an ManyToOne association, the linked entity is instantiated with null fields. The issue is not reproducing when trying to fetch OneToMany association.

Here is reported a minimum code to reproduce the problem
```java

@Entity
@Table(schema = "iot_service", name = "telemetry_data")
public class TelemetryData {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "value", nullable = false)
    private Float value;

    @Column(name = "stored_at", nullable = false)
    private Long storedAt;

    @Column(name = "probed_at", nullable = false)
    private Long probedAt;

    @JoinColumn(name = "feature_id", nullable = false)
    @ManyToOne(fetch = FetchType.LAZY, targetEntity = Feature.class)
    private Feature feature;

    public TelemetryData() {
    }

    public TelemetryData(Float value, Long storedAt, Long probedAt, Feature feature) {
        this.value = value;
        this.storedAt = storedAt;
        this.probedAt = probedAt;
        this.feature = feature;
    }

    // Getters and Setters...
}
```

```java

@Entity
@Table(schema = "iot_service", name = "features")
public class Feature {

    @Id
    @Column(name = "id")
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "feature_id", nullable = false, unique = true)
    private UUID featureId;

    @Column(name = "name", nullable = false)
    private String name;

    @Column(name = "topic_path", nullable = false, unique = true)
    private String topicPath;

    @Enumerated(value = EnumType.STRING)
    @Column(name = "feature_type", nullable = false)
    private FeatureType featureType;

    @Column(name = "is_running")
    private Boolean isRunning;

    @Column(name = "source_type")
    private SourceType sourceType;

    @Column(name = "running_reference_id")
    private String runningReferenceId;

    @OneToMany(fetch = FetchType.LAZY, targetEntity = TelemetryData.class, mappedBy = "feature")
    private List<TelemetryData> telemetryData;

    // Constructors, getters, setters...
}
```

```java

@ApplicationScoped
public class TelemetryDataRepository {

    @Inject
    Mutiny.SessionFactory rxSessionFactory;
    
    public Uni<TelemetryData> getLatestDataByFeature(Long featureId) {
        var session = this.rxSessionFactory.openSession();
        return session.createQuery(
                        "from TelemetryData t where t.feature.id = :featureId order by t.probedAt desc",
                        TelemetryData.class
                )
                .setParameter("featureId", featureId)
                .setMaxResults(1)
                .getSingleResultOrNull()
                .call(telemetryData ->
                        Mutiny.fetch(telemetryData.getFeature())
                )
                .call(ignored -> session.flush())
                .call(ignored -> session.close());
    }
}
```

Edit:
Add a sample code to retrieve data fetched by the query

```java
  this.telemetryDataRepository.getLatestDataByFeature(1L)
                .map(telemetryData -> {
                    var telemetryCalendar = Calendar.getInstance();
                    telemetryCalendar.setTimeInMillis(telemetryData.getProbedAt());
                    var telemetryUpdateMessage = new TelemetryUpdateMessage(
                            telemetryData.getFeature().getDevice().getId(), // <----- exception due null pointer exception on  feature.getDevice()
                            telemetryData.getFeature().getId(),
                            telemetryCalendar,
                            telemetryData.getValue()
                    );
                    return telemetryUpdateMessage;
                });
```

The example below is applicable also if i print the data in a logger

**Comment by gavinking** [comment]
I don't see what you do to access the fields of the `Feature`. (Note: if you're just looking in the _debugger_, you're going to be seeing the fields of the *proxy*, not the actual instance.)

**Comment by Davide-Di-Do** [comment]
> I don't see what you do to access the fields of the `Feature`. (Note: if you're just looking in the _debugger_, you're going to be seeing the fields of the _proxy_, not the actual instance.)

Sorry, I're right. I've edited my original post adding the read method of fetched data.

**Comment by DavideD** [comment]
@blafond Could you adapt this to an Hibernate Reactive test case and have a look? Unless @gavinking is already taking care of it

**Comment by blafond** [comment]
@Davide-Di-Do your `Feature` entity doesn't have a **device** field. Is that calculated in side the classe?

**Comment by DavideD** [comment]
I think the issue is in this line:
```
.call(telemetryData ->
    Mutiny.fetch(telemetryData.getFeature())
)
```

`Mutiny.fetch` is not going to initialize the association in telemetryData.
Somethin like this should work:
```
.call( telemetryData ->
    Mutiny.fetch( telemetryDatagetFeature() ).invoke( fetched -> telemetryData.setFeature( fetched ) );
) )

```

But I'm surprised you have a `NullPointerException`, you should see something like:
```
could not initialize proxy [org.hibernate.reactive.Feature#1] - no Session
```

If you could provide a testcase for Quarkus or something that with can run, it would be very helpful.

**Comment by Davide-Di-Do** [comment]
Sorry for the delay, but meanwhile some good news in my family. I've created a new project where is possibile to test the bug described [Quarkus Hibernate reactive project](https://github.com/Davide-Di-Do/hibernate-bug-testing)

**Comment by DavideD** [comment]
Thanks, I've reopened the issue and I will check the new test case later

**Comment by Davide-Di-Do** [comment]
Thanks a lot, in the project I've proposed all test cases. If need help to run the project, please contact me and I will be happy to help the resolution of the problem.

**Comment by DavideD** [comment]
I can see the error and I suspect that's something related to bytecode enhancements (something that's not enabled in our tests on Hibernate Reactive). Probably that's the reason I couldn't reproduce the issue.
I will try to figure out what's going on tomorrow.

Anyway, the solution in `getAllFeaturesWithMasterDeviceWorks` is the right approach to use for these scenario.
Except that, instead of  `.call(ignored -> rxSession.close())`, you should use `.eventually(rxSession::close)` or even better, you could use the `withSession`:
```
    public Uni<List<Feature>> getAllFeaturesWithMasterDeviceWorks() {
        return this.rxSessionFactory
		        .withSession(rxSession -> rxSession
				        .createQuery("from Feature features join fetch features.device", Feature.class)
                        .getResultList()
                );
    }
```

**Comment by Davide-Di-Do** [comment]
@DavideD thanks for advices. I see that I can enhance my code in some points. The code proposed is taken from Hibernate Reactive documentation, so can be useful update, or integrate, the documentation itself to help coding in the right way. My thoughts obvously are just suggestions and not for polemics :)

**Comment by gavinking** [comment]
@Davide-Di-Do 

> The code proposed is taken from Hibernate Reactive documentation, so can be useful update, or integrate, the documentation itself to help coding in the right way.

Wait, you found something like `.call(ignored -> rxSession.close())` somewhere in the docs?

Where? I can't seem to find it there...

**Comment by gavinking** [comment]
> Where? I can't seem to find it there...

The one place I see where we mention the `close()` method does use `eventually()` in the example.

**Comment by Davide-Di-Do** [comment]
Sorry for missunderstanding, my thought about documentation was not about session close, but instead about relations fetching

**Comment by DavideD** [comment]
Thanks for the feedback.

[In the examples](http://hibernate.org/reactive/documentation/1.1/reference/html_single/#_fetching_lazy_associations) in the documentation we use the loaded association as soon as it's retrieved. I guess we can briefly remind the reader that eagerly fetching an association via queries or fetching profiles (like you would do with JPA) is still an option.

Is there something else that you think it would be helpful to add?

Basically we are trying not to duplicate in the Hibernate Reactive documentation what's already explained by JPA or Hibernate ORM (like mapping the entities and eagerly fetching with a query).

**Comment by Davide-Di-Do** [comment]
> Basically we are trying not to duplicate in the Hibernate Reactive documentation what's already explained by JPA or Hibernate ORM (like mapping the entities and eagerly fetching with a query).

Ok, I understood. So the problem with association fetch in the query was my missing point. Excuse me for the confusion with informations.

For the part of Mutiny.fetch the problem, in perspective, persists so I think that I will continue to use the join fetch in query, but the expected result of Mutiny.fetch should be fixed at some point because in one way the fetch works, in other way not. I've not tested what happens with many-to-many or one-to-one association.

**Comment by DavideD** [comment]
Yes, this is a bug. I'm working on it.

It happens when bytecode enhancements are enabled.
Normally, one needs to enable them in the configuration but Quarkus enables them by default.

Note that in this scenario, if you already know that you always need to fetch the association, the query is the best option anyway.

**Comment by gavinking** [comment]
Rrrrrrrmmm I looked into this at one stage, but TBH I forget precisely how this works. Exactly which byte code enhancement options do you have enabled?

@Sanne probably knows more about this than me.

**Comment by gavinking** [comment]
@DavideD do you have a test case I can debug?

**Comment by DavideD** [comment]
I don't have a testcase, but if you run the `MutinyMain` example I have on my branch https://github.com/DavideD/hibernate-reactive/tree/lazy-many-to-one you will see the issue: https://github.com/hibernate/hibernate-reactive/commit/99f112dffc8bfb3e9d23856de37a1585cb41eb6b

**Comment by gavinking** [comment]
Alright, so it looks like we're going to need a reactive version of `AbstractEntityPersister.initializeEnhancedEntityUsedAsProxy()`.

**Comment by DavideD** [comment]
Thanks. 
Are you working on it?
Not a request, just checking to make sure that there aren't multiple people working on the same issue.

**Comment by gavinking** [comment]
I will try to find time. Not going to be immediate because I have a presentation to prepare this week.

**Comment by gavinking** [comment]
Fixed now: was easier than I has estimated.
